src/nvim/plines.c

```diff
@@ -907,6 +907,19 @@ int plines_m_win(win_T *wp, linenr_T first, linenr_T last, int max)
   return MIN(max, count);
 }
 
+/// Return number of window lines a physical line range will occupy.
+/// Only considers real and filler lines.
+///
+/// Mainly used for calculating scrolling offsets.
+int plines_m_win_fill(win_T *wp, linenr_T first, linenr_T last)
+{
+  int count = 0;
```

Comment:
Fast path for the common case
```suggestion
  if (!win_may_fill(wp)) {
    return last - first + 1;
  }
  int count = 0;
```